### PR TITLE
document todos and implement bytes31

### LINKED ISSUE #377: Tracking issue: remove or document all `todo!`s

### LINKED ISSUE #391: Implement bytes31 libfuncs

### PR DESCRIPTION
**Pull request by edg-l**
Fixes #377
Fixes #391

Description of the pull request changes and motivation.

## Checklist
- [x] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

**Review by fmoletta — DISMISSED** [review]
🚀

File: src/jit_runner.rs [hunk 1]
```diff
@@ -405,6 +406,7 @@ pub fn execute(
             | CoreTypeConcrete::RangeCheck(_)
             | CoreTypeConcrete::StarkNet(StarkNetTypeConcrete::System(_))
             | CoreTypeConcrete::EcOp(_)
+            | CoreTypeConcrete::Uint128MulGuarantee(_)
```
Comment (azteca1998) [inline src/jit_runner.rs:409]: `Uint128MulGuarantee` is not a builtin. The comment below may be misleading.

File: src/libfuncs/bytes31.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        registry,
+        metadata,
+        &info.branch_signatures()[0].vars[1].ty,
+    )?;
+
+    let op = entry.append_operation(arith::constant(
+        context,
+        Attribute::parse(context, &format!("{} : {}", u8::MAX, felt252_ty)).unwrap(),
+        location,
+    ));
```
Comment (azteca1998) [inline src/libfuncs/bytes31.rs:165]: Shouldn't the range be up to `2 ** 248 - 1` (or (`1 << (31 * 8)) - 1`)?

File: src/values.rs [hunk 3]
```diff
@@ -773,7 +775,7 @@ impl ValueBuilder for CoreTypeConcrete {
             CoreTypeConcrete::Sint32(_) => false,
             CoreTypeConcrete::Sint64(_) => false,
             CoreTypeConcrete::Sint128(_) => false,
-            CoreTypeConcrete::Bytes31(_) => todo!(),
+            CoreTypeConcrete::Bytes31(_) => todo!("implement whether bytes31 is complex"),
```
Comment (azteca1998) [inline src/values.rs:778]: This should be the same as a felt252.

File: src/values.rs [hunk 4]
```diff
@@ -745,8 +747,8 @@ impl ValueBuilder for CoreTypeConcrete {
             CoreTypeConcrete::Uint32(_) => false,
             CoreTypeConcrete::Uint64(_) => false,
             CoreTypeConcrete::Uint128(_) => false,
-            CoreTypeConcrete::Uint128MulGuarantee(_) => todo!(),
-            CoreTypeConcrete::NonZero(_) => todo!(),
+            CoreTypeConcrete::Uint128MulGuarantee(_) => false,
+            CoreTypeConcrete::NonZero(_) => false,
```
Comment (azteca1998) [inline src/values.rs:751]: What about `NonZero<felt252>`? We should recurse for types that have inner data mimicking its layout.
I already implemented the fix in my AOT PR and changing it here would create more merge conflicts there.

File: src/libfuncs/bytes31.rs (hunk 2 above)
Comment (edg-l) [reply src/libfuncs/bytes31.rs:165]: true, i copy pasted it from u8 and forgot to change it

**Review by azteca1998 — APPROVED** [review]
Perfect.

**Review by azteca1998 — APPROVED** [review]

**Review by igaray — APPROVED** [review]

**Pull request merged**